lib/ex_plasma/utxo.ex

```diff
@@ -66,55 +65,62 @@ defmodule ExPlasma.Utxo do
       # Create a Utxo from an Output RLP list
       iex> alias ExPlasma.Utxo
       iex> Utxo.new([<<1>>, [<<205, 193, 229, 59, 220, 116, 187, 245, 181, 247, 21, 214, 50, 125, 202, 87, 133, 226, 40, 180>>, <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>, <<13, 224, 182, 179, 167, 100, 0, 0>>]])
-      %ExPlasma.Utxo{
+      {:ok, %ExPlasma.Utxo{
         amount: 1000000000000000000,
-        blknum: 0,
+        blknum: nil,
         currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
-        oindex: 0,
+        oindex: nil,
         output_type: 1,
         owner: <<205, 193, 229, 59, 220, 116, 187, 245, 181, 247, 21, 214, 50, 125,
           202, 87, 133, 226, 40, 180>>,
-        txindex: 0
-      }
+        txindex: nil
+      }}
 
       # Create a Utxo from an Input RLP Item (encoded utxo position)
       iex> alias ExPlasma.Utxo
       iex> Utxo.new(<<233, 16, 63, 218, 0>>)
-      %ExPlasma.Utxo{
-        amount: 0,
+      {:ok, %ExPlasma.Utxo{
+        amount: nil,
         blknum: 1001,
-        currency: "0x0000000000000000000000000000000000000000",
+        currency: nil,
         oindex: 0,
-        owner: "0x0000000000000000000000000000000000000000",
+        owner: nil,
         txindex: 0
-      }
+      }}
 
       # Create a Utxo from a Utxo position.
       iex> alias ExPlasma.Utxo
       iex> pos = 2000010001
       iex> Utxo.new(pos)
-      %ExPlasma.Utxo{
-        amount: 0,
+      {:ok, %ExPlasma.Utxo{
+        amount: nil,
         blknum: 2,
-        currency: "0x0000000000000000000000000000000000000000",
+        currency: nil,
         oindex: 1,
-        owner: "0x0000000000000000000000000000000000000000",
+        owner: nil,
         txindex: 1
-      }
+      }}
   """
-  @spec new(binary() | nonempty_maybe_improper_list() | non_neg_integer()) :: __MODULE__.t()
-  def new([<<output_type>>, rest_of_output]), do: new([output_type, rest_of_output])
+  @spec new(binary() | nonempty_maybe_improper_list() | non_neg_integer()) ::
+          {:ok, __MODULE__.t()}
+  def new(data) when is_list(data), do: data |> new!() |> validate_output()
+  def new(%__MODULE__{owner: nil, currency: nil, amount: nil} = data), do: validate_input(data)
+  def new(%__MODULE__{} = data), do: validate_output(data)
+  def new(data), do: {:ok, new!(data)}
+
+  @spec new!(binary() | nonempty_maybe_improper_list() | non_neg_integer()) :: __MODULE__.t()
+  def new!([<<output_type>>, rest_of_output]), do: new!([output_type, rest_of_output])
```

Comment:
`new!` is the permissive version, but Elixir convention on `!` is different https://stackoverflow.com/questions/32449423/whats-the-meaning-of-and-syntax-in-elixir/32456168#32456168. `!` only hints about how the error can manifest (return value or being raised) not if at all.

Reply:
good one @pdobacz !

Reply:
What's a good convention to express this different behavior? I roughly follow how `decode/1` and `decode!/1` was done here: https://github.com/omisego/elixir-omg/blob/master/apps/omg/lib/omg/state/transaction.ex#L74 to circumvent the tuple(and in this specific case, also the validations)

Reply:
`def new/1`

`defp do_new/1`

Reply:
hm, but `new!` is public and `def do_new/1` might go against another convention (?) to reserve the `do_` to private worker functions.

Hmmm... I admit I'm out of ideas... `def new_dont_validate/1` :sweat_smile: ? or rearrange altogether. Like, if publicness of `new!` is only for the sake of EIP712's special thingy, maybe don't have it at all :thinking: ?